From 3750aa9bf92888df7f2524adebf5831955dbfcbb Mon Sep 17 00:00:00 2001
From: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
Date: Wed, 16 Nov 2022 20:40:50 +0530
Subject: [PATCH] wifi: ath12k: Do partner rx descriptor init in each SoC

In MLO setup, REO UMAC recieves the partner SoC Rxdma buffer. In that case,
to make the HW conversion work. Each SoC need to configure the partner SoC
Rxdma rx descriptor in the cmem configuration.

Signed-off-by: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c |  3 +++
 drivers/net/wireless/ath/ath12k/dp.c   | 17 +++++++++++++++++
 drivers/net/wireless/ath/ath12k/dp.h   |  1 +
 3 files changed, 21 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -1024,6 +1024,9 @@ static int ath12k_core_mlo_setup(struct
 			goto err_setup;
 	}
 
+	for (i = 0; i < ag->num_chip; i++)
+		ath12k_dp_partner_cc_init(ag->ab[i]);
+
 	for (i = 0; i < ag->num_hw; i++) {
 		ah = ag->ah[i];
 		if (!ah)
--- a/drivers/net/wireless/ath/ath12k/dp.c
+++ b/drivers/net/wireless/ath/ath12k/dp.c
@@ -1499,6 +1499,23 @@ static int ath12k_dp_cc_desc_init(struct
 	return 0;
 }
 
+void ath12k_dp_partner_cc_init(struct ath12k_base *ab)
+{
+	struct ath12k_hw_group *ag = ab->ag;
+	struct ath12k_base *partner_ab;
+	struct ath12k_dp *dp;
+	int i;
+
+	for (i = 0; i < ag->num_chip; i++) {
+		partner_ab = ag->ab[i];
+		if (partner_ab == ab)
+			continue;
+
+		dp = &partner_ab->dp;
+		ath12k_dp_rx_cmem_init(ab, dp);
+	}
+}
+
 static int ath12k_dp_cc_init(struct ath12k_base *ab)
 {
 	struct ath12k_dp *dp = &ab->dp;
--- a/drivers/net/wireless/ath/ath12k/dp.h
+++ b/drivers/net/wireless/ath/ath12k/dp.h
@@ -2109,6 +2109,7 @@ void ath12k_dp_vdev_tx_attach(struct ath
 void ath12k_dp_free(struct ath12k_base *ab);
 int ath12k_dp_alloc(struct ath12k_base *ab);
 void ath12k_dp_cc_config(struct ath12k_base *ab);
+void ath12k_dp_partner_cc_init(struct ath12k_base *ab);
 int ath12k_dp_pdev_alloc(struct ath12k_base *ab);
 void ath12k_dp_pdev_pre_alloc(struct ath12k_base *ab);
 void ath12k_dp_pdev_free(struct ath12k_base *ab);
